lib/rubocop/cop/rspec/capybara/specific_matcher.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      module Capybara
+        # Checks for there is a more specific matcher offered by Capybara.
+        #
+        # @example
+        #
+        #   # bad
+        #   expect(page).to have_selector('button')
+        #   expect(page).to have_no_selector('button.cls')
+        #   expect(page).to have_css('button')
+        #   expect(page).to have_no_css('a.cls', exact_text: 'foo')
+        #   expect(page).to have_css('table.cls')
+        #   expect(page).to have_css('select')
+        #
+        #   # good
+        #   expect(page).to have_button
+        #   expect(page).to have_no_button(class: 'cls')
+        #   expect(page).to have_button
+        #   expect(page).to have_no_link('foo', class: 'cls')
+        #   expect(page).to have_table(class: 'cls')
+        #   expect(page).to have_select
+        #
+        class SpecificMatcher < Base
+          MSG = 'Use `%<good_matcher>s`.'
+          RESTRICT_ON_SEND = %i[have_selector have_no_selector have_css
+                                have_no_css].freeze
+          SPECIFIC_MATCHER = {
+            'button' => 'button',
+            'a' => 'link',
+            'table' => 'table',
+            'select' => 'select'
+          }.freeze
+
+          # @!method first_argument(node)
+          def_node_matcher :first_argument, <<-PATTERN
+            (send nil? _ (str $_) ... )
+          PATTERN
+
+          def on_send(node)
+            return unless (matcher = specific_matcher(node))
+
+            add_offense(node, message: message(node, matcher))
+          end
+
+          private
+
+          def specific_matcher(node)
+            arg = first_argument(node)[/^\w+/, 0]
```

Comment:
This errors out on this:
```ruby
# real-world-rspec/administrate/spec/support/have_flash_matcher.rb
module Features
  def have_flash(text, options = {})
    options.reverse_merge!(type: :notice)
    have_css(".flash-#{options[:type]}", text: text)
  end
end
```

The first argument is a `dstr`, not `str`.

A few similar cases appear if you run it against `real-world-rspec`:
```ruby
# real-world-rspec/administrate/spec/administrate/views/fields/url/_index_spec.rb
    expect(rendered).to have_css(
      %{a[href="#{product.image_url}"]},
      text: product.image_url,
    )
```

I'd suggest adding an example to prevent regressions after fixing.

Reply:
It's fine to ignore everything apart from `str` I believe.

Reply:
Oops, I had not noticed that pattern. Thanks, I've decided to ignore the `dstr`.